EDU-3236: Update-with-Start (#3120)
temporalio/documentation: docs/develop/go/message-passing.mdx

Use the [`Client.SignalWithStartWorkflow`](https://pkg.go.dev/go.temporal.io/sdk/client#Client.SignalWithStartWorkflow) API to start a Workflow Execution (if not already running) and pass it the Signal at the same time.

> Because the Workflow Execution might not exist, this API does not take a Run Id as a parameter

Reviewer: Confirmed correct.

### Send an Update {#send-update-from-client}

:::tip Stability

> In [Pre-release](/evaluate/development-production-features/release-stages#pre-release) (API is subject to change).

Reviewer: This needs to be updated to Public Preview. We do not release docs (as a rule) before then, so it should reflect the status when this material is ready for going live.

> Minimum recommended [Temporal Server version 1.25](https://github.com/temporalio/temporal/releases/tag/v1.25.0)

Reviewer: Consider "Minimum Temporal Server version" unless there's a compelling reason otherwise to be less restrictive.

:::

An Update is a synchronous, blocking call that can change Workflow state, control its flow, and return a result.

A Client sending an Update must wait until the Server delivers the Update to a Worker.
Workers must be available and responsive.

> Therefore, provide a context with a timeout to limit the time the client waits for a response.

Reviewer: ```suggestion
Setting a timeout with a context provides a hard limit on how long a client will wait for a response.
```

If you supply only the Workflow Id (and provide an empty string as the Run Id param), the running Workflow Execution receives the Update.

You must provide a `WaitForStage` when calling `UpdateWorkflow()`.

> This parameter controls what stage the update must reach before a handle is returned to the caller.

Reviewer: ```suggestion
This parameter controls the stage the update must reach before returning a handle to the caller:

- If `WaitForStage` is set to `WorkflowUpdateStageCompleted`, the handle is returned after the Update completes.
- If `WaitForStage` is set to `WorkflowUpdateStageAccepted`, the handle is returned after the Update is accepted and the Validator is run, if it's defined.

Note: You cannot send Updates to other Workflow Executions.

The following sample demonstrates how to create a context with timeout for sending an `Update` to the Workflow:

```

Or you might consider putting the note into an admonition if important enough.

if err != nil {
    log.Fatalf("Unable to update workflow: %v", err)
}

var previousLang message.Language
err = updateHandle.Get(ctxWithTimeout, &previousLang)
if err != nil {
    log.Fatalf("Unable to get update result: %v", err)
}
```

#### Update-With-Start {#update-with-start}

> Update-With-Start lets you send an Update that checks whether an already-running Workflow with that ID exists.

Reviewer 2: Should we backlink to the encyclopedia entry here?

If it does, the Update is delivered.
If not, it starts a new Workflow Execution with the supplied ID.
Upon starting a new Workflow, it immediately processes the Update.

To create a new Update operation, use [`client.NewUpdateWithStartWorkflowOperation`](https://pkg.go.dev/go.temporal.io/sdk/client#NewUpdateWithStartWorkflowOperation).
The Workflow Id is optional - but when it is specified, it must match the one in `StartWorkflowOptions`.
It is invalid to set the Run ID since there might not be running Workflow Execution.

Use the `WithStartOperation` option to attach an Update Operation to the [`Client.ExecuteWorkflow`](https://pkg.go.dev/go.temporal.io/sdk/client#Client.ExecuteWorkflow) API call.
Not all `ExecuteWorkflow` options are allowed with Update-With-Start.
For example, specifying a Cron Schedule results in an error.

> Refer to the API documentation for further details.

Reviewer 2: Permalink?

Reviewer: Yes please

Author: I thought users could get there from the link that's there, but it does take 2 clicks -  so I added a direct link 👍

defer cancel()

updateOperation := client.NewUpdateWithStartWorkflowOperation(
    client.UpdateWorkflowOptions{
        UpdateName:   message.SetLanguageUpdate,
        WaitForStage: client.WorkflowUpdateStageCompleted,
    })

startWorkflowOptions := client.StartWorkflowOptions{
    ID:                       "some-workflow-id",
    TaskQueue:                "some-task-queue",
    WithStartOperation:       updateOperation,  // attaches the Update to the Workflow Start
	WorkflowIDConflictPolicy: enumspb.WORKFLOW_ID_CONFLICT_POLICY_USE_EXISTING,
}
we, err := c.ExecuteWorkflow(ctxWithTimeout, startWorkflowOptions, MyWorkflow)
if err != nil {
    log.Fatalf("Unable to execute workflow: %v", err)
}

updateHandle, err := updateOperation.Get(ctxWithTimeout)
if err != nil {
    log.Fatalf("Unable to obtain update handle: %v", err)
}

var previousLang message.Language
err = updateHandle.Get(ctxWithTimeout, &previousLang)
if err != nil {
    log.Fatalf("Unable to obtain update result: %v", err)
}
```

For more examples, see [Go sample for early-return pattern](https://github.com/temporalio/samples-go/tree/main/early-return).

> 

Author: I thought about adding some details about how an Update rejection for when a Workflow _is_ started, will always appear in the Workflow History. But I wasn't sure that that is sth users would find very important.

Reviewer: I think this is useful info at the point where it makes sense.

Author: Talked to Drew and we both think it's not needed.

if err != nil {
    log.Fatalf("Unable to update workflow: %v", err)
}

var previousLang message.Language

> err = updateHandle.Get(ctxWithTimeout, &previousLang)

Reviewer: Nice.

}
```

#### Update-With-Start {#update-with-start}

> Sending an Update-With-Start, if there is a running Workflow with the given Workflow Id, it will process the Update. Otherwise, a new Workflow starts and immediately processes the Update.

Reviewer: ```suggestion
Update-With-Start lets you send an Update that checks whether an already-running Workflow with that ID exists.
If it does, the Update is delivered.
If not, it starts a new Workflow Execution with the supplied ID.
Upon starting a new Workflow, it immediately processes the Update.
```

> Use the [`client.NewUpdateWithStartWorkflowOperation`](https://pkg.go.dev/go.temporal.io/sdk/client#NewUpdateWithStartWorkflowOperation) API to create a new Update Operation. The `WorkflowID` option is optional; and it is invalid to set the `RunID` option.

Reviewer: ```suggestion
To create a new Update operation, use [`client.NewUpdateWithStartWorkflowOperation`](https://pkg.go.dev/go.temporal.io/sdk/client#NewUpdateWithStartWorkflowOperation).
For most calls, the `WorkflowID` option is optional; and it is invalid to set the `RunID` option.
For `WithStartOperation`, a `WorkflowID` is required.
```

> Then, use the `WithStartOperation` option to attach the Update Operation to the [`Client.ExecuteWorkflow`](https://pkg.go.dev/go.temporal.io/sdk/client#Client.ExecuteWorkflow) API call. Note that not all `ExecuteWorkflow` options are allowed for Update-With-Start, for example specifying a `CronSchedule` with result in an error. Refer to the API documentation for details.

Reviewer: ```suggestion
Use the `WithStartOperation` option to attach an Update Operation to the [`Client.ExecuteWorkflow`](https://pkg.go.dev/go.temporal.io/sdk/client#Client.ExecuteWorkflow) API call.
Not all `ExecuteWorkflow` options are allowed with Update-With-Start.
For example, specifying a `CronSchedule` results in an error.
Refer to the API documentation for further details.
```

> The `ExecuteWorkflow` call will return once an Update result is available; or when the provided context times out. To obtain the Update result, first obtain a [`WorkflowUpdateHandle`](https://pkg.go.dev/go.temporal.io/sdk/client#WorkflowUpdateHandle) by calling `Get` on the Update operation.

Reviewer: ```suggestion
The `ExecuteWorkflow` returns once an Update result is available; or when a provided context times out.
You need a [`WorkflowUpdateHandle`](https://pkg.go.dev/go.temporal.io/sdk/client#WorkflowUpdateHandle) to retrieve a result.
Call `Get` on the Update operation.
```

> Note that an `UpdateWithStartWorkflowOperation` can only be executed once. Re-using a previously executed operation returns an error from `ExecuteWorkflow`.

Reviewer: ```suggestion
`UpdateWithStartWorkflowOperation` can only be executed once.
Re-using a previously executed operation returns an error from `ExecuteWorkflow`.
```

if err != nil {
    log.Fatalf("Unable to update workflow: %v", err)
}

var previousLang message.Language
err = updateHandle.Get(ctxWithTimeout, &previousLang)
if err != nil {
    log.Fatalf("Unable to get update result: %v", err)
}
```

#### Update-With-Start {#update-with-start}

> [Update-with-Start](/encyclopedia/workflow-message-passing#update-with-start) lets you send an Update that checks whether an already-running Workflow with that ID exists.

Reviewer: Suggestion:

```
[Update-with-Start](/encyclopedia/workflow-message-passing#update-with-start) lets you send an Update that checks whether an already-running Workflow with that ID exists:

- If the Workflow exists, the Update is processed.
- If the Workflow does not exist, a new Workflow Execution starts with the given ID, and the Update is processed immediately after.
```

If it does, the Update is processed.
If not, it starts a new Workflow Execution with the supplied ID.
Upon starting a new Workflow, it immediately processes the Update.

To create a new Update operation, use [`client.NewUpdateWithStartWorkflowOperation`](https://pkg.go.dev/go.temporal.io/sdk/client#NewUpdateWithStartWorkflowOperation).

> The Workflow Id is optional - but when it is specified, it must match the one in `StartWorkflowOptions`.

Reviewer: ```suggestion
The Workflow Id is optional. 
When specified, the Id must match the one used in `StartWorkflowOptions`.
```

> It is invalid to set the Run ID since there might not be running Workflow Execution.

Reviewer: ```suggestion
Since a running Workflow Execution may not already exist, you can't set a Run ID.
```

> Use the `WithStartOperation` option to attach an Update Operation to the [`Client.ExecuteWorkflow`](https://pkg.go.dev/go.temporal.io/sdk/client#Client.ExecuteWorkflow) API call.

Reviewer: ```suggestion
Use the `WithStartOperation` option to attach an Update Operation to the [`Client.ExecuteWorkflow`](https://pkg.go.dev/go.temporal.io/sdk/client#Client.ExecuteWorkflow) API call.

Not all `ExecuteWorkflow` options are allowed with Update-With-Start.
For example, specifying a Cron Schedule will result in an error.

For more details, refer to the [API documentation](https://pkg.go.dev/go.temporal.io/sdk/internal#StartWorkflowOptions).
```

Not all `ExecuteWorkflow` options are allowed with Update-With-Start.
For example, specifying a Cron Schedule results in an error.
Refer to the [API documentation](https://pkg.go.dev/go.temporal.io/sdk/internal#StartWorkflowOptions) for further details.

`ExecuteWorkflow` returns once the requested Update wait stage has been reached; or when a provided context times out.

> Choose [Workflow Id Conflict Policy](#workflow-id-conflict-policy) "Use Existing" coupled with an indempotent Update handler by default to allow the code to be executed again, in case of client failure.

Reviewer: ```suggestion
By default, choose the [Workflow Id Conflict Policy](#workflow-id-conflict-policy) "Use Existing" and use an idempotent Update handler.
This ensures your code can be executed again in case of a Client failure.
```

> You need a [`WorkflowUpdateHandle`](https://pkg.go.dev/go.temporal.io/sdk/client#WorkflowUpdateHandle) to retrieve a result.

Reviewer: ```suggestion
Use [`WorkflowUpdateHandle`](https://pkg.go.dev/go.temporal.io/sdk/client#WorkflowUpdateHandle) to retrieve a result from the Update.
```

Call `Get` on the Update operation.

`UpdateWithStartWorkflowOperation` can only be executed once.

> Re-using a previously executed operation returns an error from `ExecuteWorkflow`.

Reviewer: Nice.

> The following example shows the creation, configuration, and use of UpdateWithStart:

Reviewer: Also, very nice.
